util/topsql/stmtstats/aggregator.go

```diff
@@ -0,0 +1,182 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stmtstats
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// globalAggregator is global *aggregator.
+var globalAggregator atomic.Value
+
+// StatementStatsRecord is the merged StatementStatsMap with timestamp.
+type StatementStatsRecord struct {
+	Timestamp int64
+	Data      StatementStatsMap
+}
+
+// aggregator is used to collect and aggregate data from all StatementStats.
+// It is responsible for collecting data from all StatementStats, aggregating
+// them together, uploading them and regularly cleaning up the closed StatementStats.
+type aggregator struct {
+	ctx         context.Context
+	cancel      context.CancelFunc
+	statsSet    sync.Map // map[*StatementStats]struct{}
+	collectors  sync.Map // map[uint64]Collector
+	collectorID uint64
+	running     int32
+}
+
+// newAggregator creates an empty aggregator.
+func newAggregator() *aggregator {
+	return &aggregator{}
+}
+
+// run will block the current goroutine and execute the main loop of aggregator.
+func (m *aggregator) run() {
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+	atomic.StoreInt32(&m.running, 1)
+	defer func() {
+		atomic.StoreInt32(&m.running, 0)
+	}()
+	tick := time.NewTicker(time.Second)
+	defer tick.Stop()
+	for {
+		select {
+		case <-m.ctx.Done():
+			return
+		case <-tick.C:
+			m.aggregate()
+		}
+	}
+}
+
+// aggregate data from all associated StatementStats.
+// If StatementStats has been closed, collect will remove it from the map.
+func (m *aggregator) aggregate() {
+	r := StatementStatsRecord{
+		Timestamp: time.Now().Unix(),
+		Data:      StatementStatsMap{},
+	}
+	m.statsSet.Range(func(statsR, _ interface{}) bool {
+		stats := statsR.(*StatementStats)
+		if stats.Finished() {
+			m.unregister(stats)
+		}
+		r.Data.Merge(stats.Take())
+		return true
+	})
+	m.collectors.Range(func(_, c interface{}) bool {
+		c.(Collector).CollectStmtStatsRecords([]StatementStatsRecord{r})
+		return true
+	})
+}
+
+// register binds StatementStats to aggregator.
+// register is thread-safe.
+func (m *aggregator) register(stats *StatementStats) {
+	m.statsSet.Store(stats, struct{}{})
+}
+
+// unregister removes StatementStats from aggregator.
+// unregister is thread-safe.
+func (m *aggregator) unregister(stats *StatementStats) {
+	m.statsSet.Delete(stats)
+}
+
+// registerCollector binds a Collector to aggregator, collector ID will be returned.
+func (m *aggregator) registerCollector(collector Collector) uint64 {
+	id := atomic.AddUint64(&m.collectorID, 1)
+	m.collectors.Store(id, collector)
+	return id
+}
+
+// unregisterCollector removes Collector from aggregator by collector ID.
+func (m *aggregator) unregisterCollector(id uint64) {
+	m.collectors.Delete(id)
+}
```

Comment:
how about: 
```go
func (m *aggregator) registerCollector(collector Collector) {
	m.collectors.Store(collector, nil)
}

func (m *aggregator) unregisterCollector(collector Collector) {
	m.collectors.Delete(collector)
}
```

Reply:
We have this wrapper func which implemented `Collector`:
```go
// CollectorFunc implements Collector, in order to facilitate the user to
// write callback function directly.
type CollectorFunc func([]StatementStatsRecord)
```

So we can simply use:
```go
a.registerCollector(CollectorFunc(func(rs []StatementStatsRecord) {
    // do sth.
}))
```

And `CollectorFunc` cannot be used as the key of the map. So `Collector` also cannot be used as the key, unless we remove `CollectorFunc` type.

Reply:
Do you think it is necessary to keep this type? or remove it. cc @crazycs520